### PR TITLE
Enhancement/Fix: Resolve Queue Full Issue in Frame Processing

### PR DESCRIPTION
**Pull request by humayoonsiddiqui**
## Description

This pull request addresses an issue related to the queue becoming full during frame processing in the rtsp-client library. The modifications aim to enhance the stability and reliability of the library, particularly in scenarios where the queue's capacity is reached.

## Changes Made

- Introduced a timeout mechanism in the `push` method to handle queue full scenarios.
- Utilized the `offer` method to allow the system to wait for a specified time before considering the queue full.
- Implemented logic to clear the queue and interrupt the current thread if the queue size exceeded a predefined threshold.

## Version Information

This fork is based on version 2.0.6 of the original repository (alexeyvasilyev/rtsp-client-android:master). The changes made here aim to not only resolve the identified bug but also improve the overall performance of the library.

## Why Fork Instead of Wrapping Layer

The nature of the bug required modifications deep within the library's functionality, making a wrapping layer less suitable. Direct changes to the library code were necessary for an effective resolution.

## Testing

Include details on any testing performed to validate the changes made in this pull request.

### CODE REVIEW

**Review by alexeyvasilyev — CHANGES_REQUESTED** [review]
Low quality code.
Revert all intends you made. It is completely impossible to see what really has been changed.

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt [hunk 1]
```diff
@@ -57,7 +57,7 @@ open class RtspSurfaceView: SurfaceView {
     private val proxyClientListener = object: RtspClient.RtspClientListener {
 
         override fun onRtspConnecting() {
-            if (DEBUG) Log.v(TAG, "onRtspConnecting()")
+            if (DEBUG) Log.v(TAG, "onRtspConnecting..")
```
Comment (alexeyvasilyev) [inline library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt]: if (DEBUG) Log.v(TAG, "functionName()")
Should not be touched.

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/codec/AudioDecodeThread.kt [hunk 2]
```diff
@@ -1,171 +1,177 @@
...
-                16000 -> extraDataAac = extraDataAac or (0x8 shl 7)
-                22050 -> extraDataAac = extraDataAac or (0x7 shl 7)
-                24000 -> extraDataAac = extraDataAac or (0x6 shl 7)
-                32000 -> extraDataAac = extraDataAac or (0x5 shl 7)
-                44100 -> extraDataAac = extraDataAac or (0x4 shl 7)
-                48000 -> extraDataAac = extraDataAac or (0x3 shl 7)
-                64000 -> extraDataAac = extraDataAac or (0x2 shl 7)
-                88200 -> extraDataAac = extraDataAac or (0x1 shl 7)
-                96000 -> extraDataAac = extraDataAac or (0x0 shl 7)
+        companion object {
```
Comment (alexeyvasilyev) [inline library-client-rtsp/src/main/java/com/alexvas/rtsp/codec/AudioDecodeThread.kt]: Why you moved that code? Revert it.

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt [hunk 3]
```diff
@@ -30,12 +30,12 @@ open class RtspSurfaceView: SurfaceView {
     private var requestVideo = true
     private var requestAudio = true
     private var rtspThread: RtspThread? = null
-    private var videoFrameQueue = FrameQueue(60)
+    var videoFrameQueue = FrameQueue(200)
```
Comment (alexeyvasilyev) [inline library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt:30]: Why queue set to 200? Why it is not private?

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt [hunk 4]
```diff
@@ -30,12 +30,12 @@ open class RtspSurfaceView: SurfaceView {
...
     private var audioFrameQueue = FrameQueue(10)
     private var videoDecodeThread: VideoDecodeThread? = null
     private var audioDecodeThread: AudioDecodeThread? = null
-    private var surfaceWidth = 1920
-    private var surfaceHeight = 1080
+    private var surfaceWidth = 720
```
Comment (alexeyvasilyev) [inline library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt:34]: What is this?

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt [hunk 5]
```diff
@@ -81,7 +81,7 @@ open class RtspSurfaceView: SurfaceView {
                     val data = ByteArray(sps.size + pps.size)
                     sps.copyInto(data, 0, 0, sps.size)
                     pps.copyInto(data, sps.size, 0, pps.size)
-                    videoFrameQueue.push(FrameQueue.Frame(data, 0, data.size, 0))
+                    videoFrameQueue.push(FrameQueue.Frame(data, 0, data.size, 0,context))
```
Comment (alexeyvasilyev) [inline library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt:99]: "0,context" -> "0, context"

**Review by alexeyvasilyev — CHANGES_REQUESTED** [review]
I will not merge this PR. There are too many changes which completely unclear to me. Try to make several PRs isolating the problem you are trying to fix.

**Comment by FuadHepa** [comment]
First of all, thank you for this wonderful application

I hope you will make it support GU711A codec audio or help me run it in the application
Thank you

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt (hunk 3 above)
Comment (humayoonsiddiqui) [reply library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt:30]: The queue was filling up rapidly due to a combination of a long-running stream with a high frame rate and an insufficient FrameQueue size of 60. This size is adequate for streams with lower frame rates, but it becomes a bottleneck for streams that send frames more frequently.

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt (hunk 1 above)
Comment (humayoonsiddiqui) [reply library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt]: We did that for testing purpose , will revert it back.

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt (hunk 4 above)
Comment (humayoonsiddiqui) [reply library-client-rtsp/src/main/java/com/alexvas/rtsp/widget/RtspSurfaceView.kt:34]: We did that according to our application needs. But now working on it to change it dynamically.

File: library-client-rtsp/src/main/java/com/alexvas/rtsp/codec/AudioDecodeThread.kt (hunk 2 above)
Comment (humayoonsiddiqui) [reply library-client-rtsp/src/main/java/com/alexvas/rtsp/codec/AudioDecodeThread.kt]: That happened during mering, will revert it.

**Comment by humayoonsiddiqui** [comment]
> I will not merge this PR. There are too many changes which completely unclear to me. Try to make several PRs isolating the problem you are trying to fix.

I apologize that the PR is currently unclear. I understand that having too many changes can be unClear to review. I'll definitely work on splitting the changes into smaller PRs that isolate the different parts of the problem I'm trying to fix.

**Comment by FuadHepa** [comment]
First of all, thank you for this wonderful application

I hope you will make it support GU711A codec audio or help me run it in the application
Thank you

**Comment by alexeyvasilyev** [comment]
Will not merge this PR. It is still too big to understand and too many syntax errors.
Please make several smaller PRs.

**Pull request closed without merging**